python/user/bot/user_bot.py

```diff
@@ -26,29 +45,82 @@ def main():
+            top = message_heap[0]
+            if not expired(archive_state.archive_after, top.post_at):
+                logging.info("Auto-archiving caught up.")
+                break
+            logging.info(f"Auto-archiving {top.cid} posted at {top.post_at}")
+            await client.submit(exercise(top.cid, 'Archive'))
+            heapq.heappop(message_heap)
+
+        logging.info(f'started auto-archiving bot for party {party}')
+        while True:
+            try:
+                while len(message_heap) > 0 and expired(archive_state.archive_after,
+                                                        message_heap[0].post_at):
+                    top = message_heap[0]
+                    logging.info(f'archiving {Chat.Message}:{top.cid}'
+                                 f' expired after {archive_state.archive_after}s')
+                    await client.submit(exercise(top.cid, 'Archive'))
+                    heapq.heappop(message_heap)
+                    logging.info(f'{Chat.Message}:{top.cid} archived.')
+            except:
+                logging.error(f"Could not auto archive messages")
+            if len(message_heap) > 0:
+                logging.info(f'waiting for next message to archive: {message_heap[0].cid}')
+            await asyncio.sleep(bot_polling_sec)
 
     @client.ledger_created(Chat.ArchiveMessagesRequest)
     async def archive_stale_messages(event):
         logging.info(f'On {Chat.ArchiveMessagesRequest} created!')
         try:
             (_, settings_cdata) = await client.find_one(Chat.UserSettings, {'user': client.party})
-            time_thresh = datetime.datetime.now() - datetime.timedelta(
-                days=settings_cdata['archiveMessagesAfter'])
-            thresh_seconds = (time_thresh - EPOCH).total_seconds()
-            logging.info(f"time_thresh: {time_thresh}, thresh_seconds: {thresh_seconds}")
             user_messages = client.find_active(Chat.Message, {'sender': client.party})
             commands = [exercise(cid, 'Archive') for (cid, cdata) in user_messages.items() if
-                        int(cdata['postedAt']) < thresh_seconds]
+                        expired(settings_cdata['archiveMessagesAfter'], int(cdata['postedAt']))]
             logging.info(f"Will archive {len(commands)} message(s)")
             commands.append(exercise(event.cid, 'Archive'))
             await client.submit(commands)
         except:
             logging.error(f"Could not archive stale messages")
             await client.submit_exercise(event.cid, 'Archive')
 
+    @client.ledger_created(Chat.Message)
+    async def message_heapify(event):
+        message_heap = archive_state.message_heap
+        logging.info(f'On {Chat.Message} created')
+        if event.cdata['sender'] == client.party:
+            logging.info(f'New {Chat.Message} archive candidate added.')
+            heapq.heappush(message_heap,
+                           Message(int(event.cdata['postedAt']), event.cid))
```

Comment:
The implicit assumption that this bot's archival process is the only way a message might be removed from the ACS makes me a bit nervous, particularly given the fact that this code does not seem to be tolerant of the failures that would arise if it attempts to archive a contract that has already been archived by some other process.

Reply:
good point, I will add some guards, like checks by fetch by `cid` from the popped from the cache.